### ISSUE TITLE
impl IntoIterator for Attributes

### ISSUE THREAD

**Issue by hellux** [hellux/jotdown#33]
should be possible to do e.g.

```rust
for (k, v) in attrs {}
```
and
```rust
for (k, v) in &attrs {}
```

The current `Attributes::iter` function is not very idiomatic. Ideally, there should instead be implementations for `&Attributes`, `&mut Attributes` and `Attributes` with corresponding iterators: `attr::Iter`, `attr::IterMut`, `attr::IntoIter`.

**Comment by clbarnes** [comment]
Got a decent way into this but ended up struggling with lifetimes in the `IterMut`, as the `Vec::get_mut` creates a new lifetime.

```rust
/// Iterator over [Attributes] key-value pairs, in arbitrary order.
pub struct AttrIntoIter<'s> {
    items: Vec<(&'s str, AttributeValue<'s>)>,
}

impl<'s> Iterator for AttrIntoIter<'s> {
    type Item = (&'s str, AttributeValue<'s>);

    fn next(&mut self) -> Option<Self::Item> {
        self.items.pop()
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let len = self.items.len();
        (len, Some(len))
    }
}

impl<'s> IntoIterator for Attributes<'s> {
    type Item = (&'s str, AttributeValue<'s>);

    type IntoIter = AttrIntoIter<'s>;

    fn into_iter(self) -> Self::IntoIter {
        AttrIntoIter {
            items: self.0.map_or(vec![], |b| *b),
        }
    }
}

/// Iterator over references to [Attributes] key-value pairs, in arbitrary order.
pub struct AttrIter<'s> {
    attrs: &'s Attributes<'s>,
    idx: Option<usize>,
}

impl<'s> AttrIter<'s> {
    fn new(attrs: &'s Attributes<'s>) -> Self {
        let idx = attrs.len().checked_sub(1);
        Self { attrs, idx }
    }
}

impl<'s> Iterator for AttrIter<'s> {
    type Item = (&'s str, &'s AttributeValue<'s>);

    fn next(&mut self) -> Option<Self::Item> {
        self.idx = self.idx?.checked_sub(1);
        let i = self.idx?;
        self.attrs.0.as_ref().unwrap().get(i).map(|a| (a.0, &a.1))
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let len = self.idx.map_or(0, |i| i + 1);
        (len, Some(len))
    }
}

impl<'s> IntoIterator for &'s Attributes<'s> {
    type Item = (&'s str, &'s AttributeValue<'s>);

    type IntoIter = AttrIter<'s>;

    fn into_iter(self) -> Self::IntoIter {
        AttrIter::new(self)
    }
}

/// Iterator over mutable references to [Attributes] key-value pairs, in arbitrary order.
pub struct AttrIterMut<'s> {
    attrs: &'s mut Attributes<'s>,
    idx: Option<usize>,
}

impl<'s> AttrIterMut<'s> {
    fn new(attrs: &'s mut Attributes<'s>) -> Self {
        let idx = Some(attrs.len());
        Self { attrs, idx }
    }
}
```

**Comment by hellux** [comment]
closed by #45
